### DISCUSSION TITLE
Draft mode doesn't bypass fetch cache

### DISCUSSION THREAD

**Discussion by tom-sherman** [vercel/next.js#51302] [App Router] [answered]
I expected that if I have draft mode enabled, I would bypass the fetch cache in production. This isn't the case in practice. Draft mode seems to have no effect at all to the fetch cache.

My page looks like this:

```tsx
import { draftMode } from "next/headers";

export const dynamic = "force-dynamic";

export default async function Page() {
  const res = await fetch(
    "http://worldtimeapi.org/api/timezone/Europe/London",
    {
      next: {
        revalidate: 10,
      },
    }
  );
  const data = await res.json();
  const { isEnabled } = draftMode();

  return (
    <>
      <h1>{data.datetime}</h1>
      {isEnabled ? <p>Draft mode is enabled</p> : <p>Draft mode is disabled</p>}
      <form action={isEnabled ? "/disable-draft-mode" : "/enable-draft-mode"}>
        <button>{isEnabled ? "Disable" : "Enable"} draft mode</button>
      </form>
    </>
  );
}
```

In this example I'd expect to see the same timestamp displayed when refreshing in 10s windows, but when draft mode is enabled i'd expect to bypass this cache and see the dynamic fetch result.

The actual behaviour is that draft mode doesn't make any difference.

**Comment by tom-sherman** [comment, marked as answer]
The solution right now is to bypass the cache when in draft mode:

```ts
const { isEnabled } = draftMode();
const res = await fetch(
  "http://worldtimeapi.org/api/timezone/Europe/London",
  {
    cache: isEnabled ? "no-cache" : undefined,
    next: {
      revalidate: isEnabled ? undefined : 10,
    },
  }
);
```

> [!NOTE]
> You must disable revalidation as well by setting to `undefined` otherwise it overrides `cache: "no-cache"`

This might be the default in the future: https://twitter.com/cramforce/status/1669024161396846596

**Reply by WPaczula** [reply]
Huge thanks, I was wondering if I'm doing something wrong here for some time now 👍

**Reply by rgembalik** [reply]
Unfortunately, this is not a solution that will work for all. If you are working with an SDK package which wraps around fetch without giving you an option to override all fetch params, you will still be unable to bypass the fetch cache. I'd gladly see it happen by default or have it as an option to next.config which allows us to set the fetch cache (or all caches) behavior when the draft mode is enabled.

**Reply by tom-sherman** [reply]
Touched on this in the other thread: https://github.com/vercel/next.js/discussions/51302#discussioncomment-6872141

**Reply by mareksuscak** [reply]
This should be in the documentation for Draft mode in big red font. Incredibly counterintuitive. You'd just assume that if there's a special draft mode, that it disables all the caches automatically. Just lost a few hours to this issue.

**Comment by joakim-roos** [comment]
Hi, how would we do this with Route Segment Config using a third-party package for fetching? 

When in draftMode i'd like to disable cache altogether:

```js
import { cache } from 'react';
import { draftMode } from 'next/headers';

import {
  getStoryblokApi,
  ISbStoriesParams,
  StoryblokComponent,
} from '@storyblok/react/rsc';

import 'server-only';
import { StoryblokProvider } from '@/lib/storyblok/storyblok-service';
import { notFound } from 'next/navigation';

// when draftMode 0, else 30
export const revalidate = 0;

export default async function WithoutEditingPage() {
  const isEnabled = draftMode().isEnabled;
  const { data } = await getPage({ draftMode: isEnabled });

  if (!data) return notFound();

  return (
    <>
      <StoryblokComponent blok={data.story.content} />
      <StoryblokProvider
        originalStory={data.story}
        draftMode={isEnabled}
        locale={'en'}
      />
    </>
  );
}

const getPage = cache(async ({ draftMode = false }: { draftMode: boolean }) => {
  const settings: ISbStoriesParams = {
    version: draftMode ? 'draft' : 'published',
  };

  const api = getStoryblokApi();
  return api.get(`cdn/stories/without-editing`, settings);
});
```

**Reply by tom-sherman** [reply]
You can't. It's well known that these third party libraries that use `fetch` under the hood without any opportunity to change the fetch settings are hard to work with in Next.js

The best course of action would be to explain this use case to the library and get them to allow these settings to be customised.

**Reply by joakim-roos** [reply]
@tom-sherman Thanks Tom, we did a swift decision right now to not use the third-party package for fetching but implement our own instead.

**Reply by karlandin** [reply]
Thank you both! This thread really helped me to understand why my pages wasn't always updated as I would have expected in draft mode. And it led me to find a solution for overriding fetch/cache params in Storyblok. I thought I might share my solution as I see you used storyblok as well @joakim-roos. 

```
storyblokInit({
  accessToken: "<token>",
  apiOptions: {
    fetch: (input, init) => {
      const { isEnabled } = draftMode();

      // @ts-ignore
      return fetch(input, {
        ...init,
        cache: isEnabled ? "no-cache" : undefined,
        next: {
          revalidate: isEnabled ? undefined : 10,
        },
      });
    },
  },
  use: [apiPlugin],
});
```

Just a disclaimer, I'm new to both Next app router and Storyblok, and I have not tested this much yet, so I am not sure this is the best/correct solution.

**Reply by Jdruwe** [reply]
@karlandin I just tried this but for me it's still failing. I still see outdated data, anything changed in the meanwhile?

**Comment by Jdruwe** [comment]
> @karlandin I just tried this but for me it's still failing. I still see outdated data, anything changed in the meanwhile?

Found it, they changed it last week :D You now can pass fetchOptions to the methods themselves:

```    
return storyblokApi.get(`cdn/stories/home`, sbParams, {
        cache: draftMode().isEnabled ? "no-cache" : undefined,
        next: {
            revalidate: draftMode().isEnabled ? undefined : 3600,
        },
    });
```
